File: 누적 합, 세그먼트 트리/(심화)G2.10800 컬러볼/조연주/main.cc
// 8332 KB, 88 ms
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

struct ball {
	int c, s, idx; // 색, 크기, 순서
};

vector <ball> v;

int N;
int C[200001]; // 색깔별 합
int S[2001]; // 크기별 합
int ans[200001]; // 정답 배열

// 정렬 기준
bool compare(ball& a, ball& b) {
	if (a.s == b.s) {
		return a.c < b.c;
	}
	return a.s < b.s;
}

int main()
{
	ios::sync_with_stdio(false);
	cin.tie(NULL); cout.tie(NULL);
    
	cin >> N; // 공 개수

	// 공 정보 입력받기
	for (int i = 0; i < N; i++) {
		int c, s;
		cin >> c >> s;
		v.push_back({ c,s,i });
	}

	// 정렬
	sort(v.begin(), v.end(),compare);

	// 누적합
	int sum = 0;

	for (int i = 0; i < N; i++) {
		int color = v[i].c;
		int size = v[i].s;
		int idx = v[i].idx;

		sum += size;
		C[color] += size;
		S[size] += size;

		// 이전 공과 크기, 색이 모두 같은 경우
		if (i != 0 && color == v[i - 1].c && size == v[i - 1].s) {
			ans[idx] = ans[v[i - 1].idx];
		}
		// 그렇지 않은 경우
		else {
			// 현재까지의 합 - 같은 색 합 - 같은 크기 합 - 자신의 크기
			ans[idx] = sum - C[color] - S[size] + size;
		}
	}

	// 출력
	for (int i = 0; i < N; i++) {
		cout << ans[i] << "\n";
	}

	return 0;
}

File: 누적 합, 세그먼트 트리/(심화)G2.10800 컬러볼/조연주/Main.java
// 85740 KB, 1260ms
import java.io.*;
import java.util.*;

class Ball implements Comparable<Ball>{
    int c,s,idx;
    public Ball(int c, int s, int idx){
        this.c = c;
        this.s = s;
        this.idx = idx;
    }

    @Override
    public int compareTo(Ball o) {
        if(o.s == this.s){
            return this.c - o.c;
        }
        return this.s - o.s;
    }
}

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static StringTokenizer st;
    static int N;
    static Ball[] balls;
    static int[] C,ans;
    static int[] S = new int[2001];

    static void prefix() throws IOException {
        int sum = 0;

        for(int i=0; i<N; i++){
            int color = balls[i].c;
            int size = balls[i].s;
            int idx = balls[i].idx;

            // 누적합
            sum += size;
            C[color] += size;
            S[size] += size;

            if(i!=0 && color == balls[i-1].c && size == balls[i-1].s){
                ans[idx] = ans[balls[i-1].idx];
            }
            else{
                ans[idx] = sum - C[color] - S[size] + size;
            }
        }
    }

    static void get() throws IOException {
        N = input(); // 공 개수
        balls = new Ball[N];
        C = new int[N+1];
        ans = new int[N];

        // 입력받기
        for(int i=0; i<N; i++){
            st = new StringTokenizer(br.readLine());
            int c = input(st);
            int s = input(st);
            balls[i] = new Ball(c,s,i);
        }

        // 정렬
        Arrays.sort(balls);
    }

    static void print() throws IOException {
        StringBuilder sb = new StringBuilder();
        for(int i=0; i<N; i++){
            sb.append(ans[i]).append("\n");
        }
        bw.write(sb.toString());
        bw.flush();
        bw.close();
    }

    public static void main(String[] args) throws IOException {
        get();
        prefix();
        print();
    }


    static int input() throws IOException {
        return Integer.parseInt(br.readLine());
    }

    static int input(StringTokenizer st) {
        return Integer.parseInt(st.nextToken());
    }

}
